l1-contracts/src/core/libraries/decoders/TxsDecoder.sol

```diff
@@ -516,6 +517,36 @@ library TxsDecoder {
     return _leafs[0];
   }
 
+  /**
+   * @notice Computes the root for a binary unbalanced Merkle-tree given the leaves and subtree sizes.
+   * @dev Filled in greedily with subtrees. Useful for txsEffectHash and outHash tree.
+   * @param _leaves - The 32 bytes leafs to build the tree of.
+   * @return The root of the Merkle tree.
+   */
+  function computeUnbalancedRoot(bytes32[] memory _leaves) internal pure returns (bytes32) {
+    // e.g. an unbalanced tree of 7 txs will contain subtrees of 4, 2, and 1 tx(s)
+    // Encoded as 000400020001
+    bytes memory subtreeSizes = MerkleLib.computeSubtreeSizes(_leaves.length);
+    // We collect the roots of each subtree
+    bytes32 root;
+    uint256 currentSubtreeSize;
+    uint256 processedLeaves = 0;
+    // We must calculate the smaller rightmost subtrees first, hence working backwards through the leaves/sizes arr
+    for (uint256 j = subtreeSizes.length; j > 0; j -= 2) {
```

Comment:
Looks like the subtreeSizes are simply the non-zero bits of the binary form of numTxs.
Consider doing the following so that we don't have to allocate some memory for the sizes:
```
uint256 currentSubtreeSize = 1;
uint256 numTxs = _leaves.length;
while (currentSubtreeSize <= numTxs) {
  if (!(currentSubtreeSize & numTxs)) {
    continue;
  }
  // Do the thing.
  currentSubtreeSize <<= 1;
}
```
(☝️ Might contain bug and incorrect syntax)